proposals/p2015.md

```diff
@@ -0,0 +1,270 @@
+-   This does not prescribe an official grammar for parsing these types
+-   This proposal does not address other, non-multiple of 8 bit sizes, such as
+    those used in a bit field
+
+## Details
+
+### Syntax
+
+The syntax for a two's complement signed integer, the unsigned integer, and the
+floating-point number corresponds to a lowercase 'i', 'u', or 'f' character,
+respectively, indicating the type followed by a numeric value specifying the
+width.
+
+As a regular expression, this can be illustrated as:
+
+```re
+([iuf])([1-9][0-9]*)
+```
+
+Capture group 1 indicates either an 'i' for a two's complement signed integer
+type, a 'u' for an unsigned integer type, or an 'f' for a floating-point number
+type. Capture group 2 specifies the width in bits. Note that this bit width is
+restricted to a multiple of 8.
+
+Examples of this syntax include:
+
+-   `i16` - A 16-bit two's complement signed integer type
+-   `u32` - A 32-bit unsigned integer type
+-   `f64` - A 64-bit IEEE 754 floating-point number type
+
+### Usage
+
+```carbon
+package sample api;
+
+fn Sum(x: i32, y: i32) -> i32 {
+  return x + y;
+}
+
+fn Main() -> i32 {
+  return Sum(4, 2);
+}
+```
+
+In the above example, `Sum` has parameters `x` and `y`, each of which is typed
+as a 32-bit two's complement signed integer. `Main` then returns the output of
+`Sum` as a 32-bit two's complement signed integer.
+
+### Syntax examples and considerations
+
+These examples are not meant to be formal mappings. They are only used to
+illustrate to the reader how the numeric type syntax may connect to the
+underlying language semantics in future numeric type additions.
+
+#### Syntax example use in common fixed-size numeric types
+
+[As proposed by @chandlerc](https://github.com/carbon-language/carbon-lang/issues/543#issuecomment-845620894),
+our heuristic for type qualification here will be "Any truly pervasive
+non-aggregate/composite/compound type that is also fundamentally
+size-parameterized."
+
+These fixed-width sizes can naturally cover the common register widths of 8-,
+16-, 32-, 64-, and 128-bits, as seen in other languages.
+
+Which expands to:
+
+-   `i8`, `i16`, `i32`, `i64`, and `i128` keyword aliases for
+    `Carbon.Int(N={8,16,32,64,128})`
+-   `u8`, `u16`, `u32`, `u64`, and `u128` keyword aliases for
+    `Carbon.UInt(N={8,16,32,64,128})`
+-   `f16`, `f32`, and `f64` keyword aliases for `Carbon.Float(N={16,32,64})`
+
+Additionally, other type sizes may be available, depending on the platform (for
+example, [x86](https://en.wikipedia.org/wiki/X86)), such as `f80`, `f128`, or
+`f256`.
+
+#### Syntax example use in arbitrary fixed-size numeric types
+
+In the more general case, this syntax can express N-width types.
+
+-   `iN` for `Carbon.Int(N)`, An N-width two's complement signed integer type
+-   `uN` for `Carbon.UInt(N)`, An N-width unsigned integer type
+-   `fN` for `Carbon.Float(N)`, An N-width IEEE 754 binary floating-point number
+    type
```

Comment:
I'm not sure what these sections are adding, particularly now that there are not differences between "common" and "arbitrary" sizes.

Reply:
Yeah, that's fair, and I agree. I'm going to remove both of these sections.